etcdserver/util.go

```diff
@@ -117,11 +145,30 @@ func warnOfExpensiveGenericRequest(lg *zap.Logger, now time.Time, stringer fmt.S
 				zap.Duration("took", d),
 				zap.Duration("expected-duration", warnApplyDuration),
 				zap.String("prefix", prefix),
-				zap.String("request", stringer.String()),
+				zap.String("request", reqStringer.String()),
+				zap.String("response", resp),
+				zap.Error(err),
 			)
 		} else {
-			plog.Warningf("%srequest %q took too long (%v) to execute", prefix, stringer.String(), d)
+			var result string
+			if err != nil {
+				result = fmt.Sprintf("<error:%v>", err)
+			} else {
+				result = resp
+			}
+			plog.Warningf("%srequest %q with result %q took too long (%v) to execute", prefix, reqStringer.String(), result, d)
 		}
 		slowApplies.Inc()
 	}
 }
+
+// safeSize attempt to calculate size using proto.Size, but recovers from panics and returns "proto_size_failure" in those cases.
+// TODO: Remove once https://github.com/golang/protobuf/issues/631 is resolved.
+func safeSize(msg proto.Message) (size string) {
+	defer func() {
+		if r := recover(); r != nil {
+			size = "proto_size_failure"
+		}
+	}()
```

Comment:
Would this work?

```go
import "github.com/golang/protobuf/proto"

ms, ok := msg.(proto.Marshaler)
if ms == nil || !ok {
  return "proto_size_failure"
}
```

Reply:
Unfortunately no, I just tried it and got the same error.

Reply:
I see. Then, let's use `recover` for now, and revert it back to upstream `proto.Size` later.

Reply:
Sounds good. Too bad we can't have a clean PR for this improvement. I'll track the upstream issue and propose a PR if needed.